server/src/main/java/io/crate/action/sql/Cursor.java

```diff
@@ -226,6 +224,14 @@ private void triggerConsumer(RowConsumer consumer, BatchIterator<Row> fullResult
         }
     }
 
+    // When cursorPosition + count >= rows.size() + 1, last row exceeded and
+    // next backwards movement must include the last row
+    private void setMaxCursorPosition() {
```

Comment:
```suggestion
    private void resetCursorToMaxBufferedRowsPlus1() {
```

Or `setPositionToBufferedTail`?
Not sure about it. I was initially confused why it's okay to jump ahead, until I realized that `rows()` is only contained the already _seen_ stuff.

Is this safe if moving foward, backward, forward again? (where the `rows` buffer exceeds the cursor position?)

Reply:
I have played around a lot, and yes it's safe, if there are more rows to be consumed and fill in the buffer, then cursorPosition is correct rows.size()+1 (we use rows.get(cursorPosition -1), more specifically `sublist()`), if it's exhausted we are past the last row and when we scroll backwards we include that last row.

Reply:
```
cr> begin;                                                                                                                                                                                                       
BEGIN OK, 0 rows affected  (0.001 sec)
cr> DECLARE cur SCROLL CURSOR FOR SELECT * FROM generate_series(1,5,1) ORDER BY 1;                                                                                                                               
DECLARE OK, 0 rows affected  (0.013 sec)
cr> fetch forward 4 from cur;                                                                                                                                                                                    
+-----------------+
| generate_series |
+-----------------+
|               1 |
|               2 |
|               3 |
|               4 |
+-----------------+
FETCH 4 rows in set (0.002 sec)
cr> fetch backward 3 from cur;                                                                                                                                                                                   
+-----------------+
| generate_series |
+-----------------+
|               3 |
|               2 |
|               1 |
+-----------------+
FETCH 3 rows in set (0.001 sec)
cr> fetch forward 2 from cur;                                                                                                                                                                                    
+-----------------+
| generate_series |
+-----------------+
|               2 |
|               3 |
+-----------------+
FETCH 2 rows in set (0.001 sec)
```